File: bindings/fsharp/src/Proven/SafeColor.fs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe color space operations via P/Invoke to libproven.
/// All computation is performed in verified Idris 2 code via the Zig FFI bridge.
/// This module contains ONLY marshaling logic.
namespace Proven

module SafeColor =

    /// Parse hex color string (#RRGGBB or #RGB).
    /// Returns the RGBColor struct on success, or None on failure.
    let parseHex (hex: string) : RGBColor option =
        let bytes = toUtf8 hex
        let result = FFI.proven_color_parse_hex(bytes, unativeint bytes.Length)
        if result.Status = 0 then Some result.Color
        else None

    /// Convert RGB color to HSL color space.
    let rgbToHsl (rgb: RGBColor) : HSLColor =
        FFI.proven_color_rgb_to_hsl(rgb)

    /// Format RGB color as hex string (e.g., "#ff0000").
    let toHex (rgb: RGBColor) : string option =
        FFI.proven_color_to_hex(rgb) |> stringResultToOption

File: bindings/nim/src/proven/safe_color.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe color operations (RGB, HSL, hex).
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

type
  Rgb* = object
    ## RGB color with components in [0, 255].
    r*: uint8
    g*: uint8
    b*: uint8

  Hsl* = object
    ## HSL color (hue in [0, 360), saturation and lightness in [0, 1]).
    h*: float64
    s*: float64
    l*: float64

proc parseHex*(hex: string): Option[Rgb] =
  ## Parse a hex color string (#RRGGBB, #RGB).
  if hex.len == 0:
    return none(Rgb)
  let res = provenColorParseHex(unsafeAddr hex[0], csize_t(hex.len))
  if res.status == PROVEN_OK:
    return some(Rgb(r: res.color.r, g: res.color.g, b: res.color.b))
  none(Rgb)

proc rgbToHsl*(c: Rgb): Hsl =
  ## Convert RGB to HSL.
  let ffiRgb = RGBColor(r: c.r, g: c.g, b: c.b)
  let res = provenColorRgbToHsl(ffiRgb)
  Hsl(h: res.h, s: res.s, l: res.l)

proc toHex*(c: Rgb): Option[string] =
  ## Convert RGB to hex string (#rrggbb).
  let ffiRgb = RGBColor(r: c.r, g: c.g, b: c.b)
  let res = provenColorToHex(ffiRgb)
  if res.status == PROVEN_OK and res.value != nil:
    let hex = $res.value
    provenFreeString(res.value)
    return some(hex)
  none(string)
